File: Difficulty: Basic/Smaller and Larger/smaller-and-larger.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

// User function template for C++
class Solution {
  public:
    vector<int> getMoreAndLess(vector<int> &arr, int target) {
        int n=arr.size();
        int lessEqual=0,greaterEqual=0;
        for(int i=0;i<n;i++){
            if(arr[i]<=target) lessEqual++;
            if(arr[i]>=target) greaterEqual++;
        }
        return {lessEqual, greaterEqual};
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    cin.ignore(); // To ignore the newline character after the integer input
    while (t--) {
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        int x;
        cin >> x;
        cin.ignore();
        Solution ob;
        vector<int> ans = ob.getMoreAndLess(arr, x);
        cout << ans[0] << " " << ans[1] << "\n~\n";
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Basic/Smaller and Larger/smaller-and-larger.java
// User function Template for Java

class Solution {
    int[] getMoreAndLess(int[] arr, int target) {
        int n = arr.length;

        // lower_bound: first index with value >= target
        int lb = lowerBound(arr, target);

        // upper_bound: first index with value > target
        int ub = upperBound(arr, target);

        int lessEqual = ub;       // count of <= target
        int greaterEqual = n - lb; // count of >= target

        return new int[]{lessEqual, greaterEqual};
    }

    // First index where arr[i] >= x
    private int lowerBound(int[] arr, int x) {
        int l = 0, r = arr.length;
        while (l < r) {
            int m = l + (r - l) / 2;
            if (arr[m] < x) l = m + 1;
            else r = m;
        }
        return l;
    }

    // First index where arr[i] > x
    private int upperBound(int[] arr, int x) {
        int l = 0, r = arr.length;
        while (l < r) {
            int m = l + (r - l) / 2;
            if (arr[m] <= x) l = m + 1;
            else r = m;
        }
        return l;
    }
}
